### PR TITLE
feat (executorIMAP)

### PR DESCRIPTION
**Pull request by yesnault**

### CODE REVIEW

File: executors/imap/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+testcases:
+- name: TestCase IMAP
+  steps:
+  - type: imap
+    imaphost: yourimaphost
+    imapuser: yourimapuser
+    imappassword: "yourimappassword"
+    mbox: INBOX
+    mboxifsuccess: mailsMatches
+    searchsubject: Title of mail
```
Comment (fsamin) [inline executors/imap/README.md]: with regexpr ?

File: executors/imap/decode.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		idx := strings.Index(date, "(") // remove (TS)
+		if idx > 0 {
+			date = date[0:idx]
+		}
+		date = strings.Trim(date, " ")
+		t, errt := time.Parse("Mon, 2 Jan 2006 15:04:05 -0700", date)
+		if errt == nil {
+			tm.Date = t
+		} else {
+			t2, errt2 := time.Parse("Fri, 17 Feb 2017 04:07:29 +0000 (UTC)", date)
```
Comment (fsamin) [inline executors/imap/decode.go]: what is this date ?

File: executors/imap/decode.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		tm.Body = string(bodya)
+		return tm, nil
+	} else if len(params) > 0 {
+		l.Debugf("Error while read body:%s", errra.Error())
+		r := bytes.NewReader(body)
+		mr := multipart.NewReader(r, params["boundary"])
+		for {
+			p, err := mr.NextPart()
+			if err == io.EOF {
+				l.Debugf("errA:%s", err)
```
Comment (fsamin) [inline executors/imap/decode.go]: missing context in error log

File: executors/imap/decode.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+				continue
+			}
+			if err != nil {
+				l.Debugf("--------> errB:%s", err)
```
Comment (fsamin) [inline executors/imap/decode.go]: missing context in error log

File: executors/imap/decode.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+				break
+			}
+			slurp, err := ioutil.ReadAll(p)
+			if err != nil {
+				l.Debugf("errC:%s", err)
```
Comment (fsamin) [inline executors/imap/decode.go]: missing context in error log

File: executors/imap/decode.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+			break
+		}
+		l.Debugf("stepD")
+	}
+
+	if tm.Body == "" {
+		l.Debugf("EmptyBody, take body")
+		tm.Body = string(bodya)
+	} else {
+		l.Debugf("stepF, tm.Body is ok")
```
Comment (fsamin) [inline executors/imap/decode.go]: useless log

File: executors/imap/executor.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	return &Executor{}
+}
+
+// Executor represents a Test Exec
+type Executor struct {
+	IMAPHost      string `json:"imaphost,omitempty" yaml:"imaphost,omitempty"`
+	IMAPUser      string `json:"imapuser,omitempty" yaml:"imapuser,omitempty"`
+	IMAPPassword  string `json:"imappassword,omitempty" yaml:"imappassword,omitempty"`
+	MBox          string `json:"mbox,omitempty" yaml:"mbox,omitempty"`
+	MBoxIfSuccess string `json:"mboxifsuccess,omitempty" yaml:"mboxifsuccess,omitempty"`
```
Comment (fsamin) [inline executors/imap/executor.go]: OnSuccess

File: executors/imap/executor.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+// GetDefaultAssertions return default assertions for type exec
+func (Executor) GetDefaultAssertions() *venom.StepAssertions {
+	return &venom.StepAssertions{Assertions: []string{"result.err ShouldNotExist"}}
+}
+
+// Run execute TestStep of type exec
+func (Executor) Run(ctx context.Context, l *log.Entry, aliases venom.Aliases, step venom.TestStep) (venom.ExecutorResult, error) {
+
```
Comment (fsamin) [inline executors/imap/executor.go]: remove empty line

File: executors/imap/executor.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+	elapsed := time.Since(start)
+	result.TimeSeconds = elapsed.Seconds()
+	result.TimeHuman = fmt.Sprintf("%s", elapsed)
+	result.Executor.IMAPPassword = "****hide****" // do not output password
+
+	return dump.ToMap(result, dump.WithDefaultLowerCaseFormatter())
+}
+
+func (e *Executor) getMail(l *log.Entry) (*Mail, error) {
+
```
Comment (fsamin) [inline executors/imap/executor.go]: remove empty line

File: executors/imap/executor.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+		return strings.Contains(m.From, e.SearchFrom) ||
+			strings.Contains(m.Subject, e.SearchSubject)
+	}
+	if e.SearchFrom != "" {
+		return strings.Contains(m.From, e.SearchFrom)
+	}
+	if e.SearchSubject != "" {
+		return strings.Contains(m.Subject, e.SearchSubject)
+	}
+
```
Comment (fsamin) [inline executors/imap/executor.go]: should manage regular expression in all searches

File: executors/imap/executor.go [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+	seq.AddNum(m.UID)
+
+	if _, err := c.UIDMove(seq, mbox); err != nil {
+		return fmt.Errorf("Error while move msg to %s, err:%s", mbox, err.Error())
+	}
+	return nil
+}
+
+func connect(host, imapUsername, imapPassword string) (*imap.Client, error) {
+
```
Comment (fsamin) [inline executors/imap/executor.go]: remove empty line

File: executors/imap/executor.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	c, errd := imap.DialTLS(host+":993", nil)
```
Comment (fsamin) [inline executors/imap/executor.go]: port should be an option, or should be appended only if `host` doesn't contain any `:`

File: executors/smtp/executor.go [hunk 13]
```diff
@@ -61,20 +61,20 @@ func (Executor) Run(ctx context.Context, l *log.Entry, aliases venom.Aliases, st
...
-	errs := t.sendEmail()
+	errs := t.sendEmail(l)
 	if errs != nil {
 		result.Err = errs.Error()
 	}
 
 	elapsed := time.Since(start)
 	result.TimeSeconds = elapsed.Seconds()
 	result.TimeHuman = fmt.Sprintf("%s", elapsed)
+	result.Executor.Password = "****hide****" // do not output password
```
Comment (fsamin) [inline executors/smtp/executor.go]: hidden

File: executors/imap/decode.go (hunk 2 above)
Comment (yesnault) [inline executors/imap/decode.go]: removed, unused.

File: executors/imap/decode.go (hunk 3 above)
Comment (yesnault) [inline executors/imap/decode.go]: fixed

File: executors/imap/decode.go (hunk 4 above)
Comment (yesnault) [inline executors/imap/decode.go]: fixed

File: executors/imap/decode.go (hunk 5 above)
Comment (yesnault) [inline executors/imap/decode.go]: fixed

File: executors/imap/decode.go (hunk 6 above)
Comment (yesnault) [inline executors/imap/decode.go]: fixed

File: executors/imap/executor.go (hunk 7 above)
Comment (yesnault) [inline executors/imap/executor.go]: fixed

File: executors/imap/executor.go (hunk 9 above)
Comment (yesnault) [inline executors/imap/executor.go]: done

File: executors/smtp/executor.go (hunk 13 above)
Comment (yesnault) [inline executors/smtp/executor.go]: done

File: executors/imap/executor.go [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+	count, err := queryCount(c, box)
+	if err != nil {
+		disconnect(c)
+		return nil, fmt.Errorf("Error while queryCount:%s", err.Error())
+	}
+
+	l.Debugf("count messages:%d", count)
+
+	if count == 0 {
+		return nil, fmt.Errorf("No message to fetch")
```
Comment (sguiheux) [inline executors/imap/executor.go]: 1 argument on fmt.Errorf =>  errors.NEw(string)

File: executors/imap/executor.go (hunk 10 above)
Comment (yesnault) [inline executors/imap/executor.go]: yep, next commit.

File: executors/imap/executor.go [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+	return dump.ToMap(result, dump.WithDefaultLowerCaseFormatter())
+}
+
+func (e *Executor) getMail(l *log.Entry) (*Mail, error) {
+
+	if e.SearchFrom == "" && e.SearchSubject == "" {
+		return nil, fmt.Errorf("You have to use searchfrom and/or searchsubject")
+	}
+
+	c, errc := connect(e.IMAPHost, e.IMAPUser, e.IMAPPassword)
```
Comment (sguiheux) [inline executors/imap/executor.go]: defer c.Close(false) here instead of disconnect ?

Comment (yesnault) [inline executors/imap/executor.go]: fixed

File: executors/imap/README.md (hunk 1 above)
Comment (yesnault) [inline executors/imap/README.md]: done.

File: executors/imap/executor.go (hunk 10 above)
Comment (yesnault) [inline executors/imap/executor.go]: done

**Pull request merged**